### PR TITLE
Dsl parser datatypes

### PR DESCRIPTION
**Pull request by konradxyz**

### CODE REVIEW

File: dsl_parser/elements/data_types.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        raise exceptions.DSLParsingElementMatchException(
+            39,
+            unknown_type_err)
+    types_classes[current_type].requires[
+        types_classes[required_type]] = [Value(required_type)]
+
+
+class DataTypes(Element):
+    schema = Dict(type=DataType)
+    provides = ['data_types']
```
Comment (simon-bar) [inline dsl_parser/elements/data_types.py]: Is this still needed?

File: dsl_parser/elements/data_types.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+from dsl_parser.elements import properties_utils
+from dsl_parser.framework.elements import Element, Dict, Leaf, DictElement
+from dsl_parser.framework.parser import parse
+from dsl_parser.framework.requirements import Value
+
+
+_DERIVED_FROM = 'derived_from'
+
+
+class StringElement(Element):
```
Comment (simon-bar) [inline dsl_parser/elements/data_types.py]: Perhaps move it to utils?

File: dsl_parser/elements/data_types.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+class DataTypes(Element):
+    schema = Dict(type=DataType)
+    provides = ['data_types']
+
+    def parse(self):
+        datatypes = self.build_dict_result()
+
+        class DataTypesInternal(DictElement):
+            schema = {}
+            requires = {}
```
Comment (simon-bar) [inline dsl_parser/elements/data_types.py]: Can we delete this line?

File: dsl_parser/elements/data_types.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            types_internal[type_name] = DataTypeInternal
+            DataTypesInternal.schema[type_name] = DataTypeInternal
+
+        for type_name, type_schema in datatypes.iteritems():
+            if type_schema.get(_DERIVED_FROM):
+                err_msg = 'Type {0} derives from unknown type {1}'.format(
+                    type_name,
+                    type_schema[_DERIVED_FROM])
+                _add_requirement_if_exists(type_name,
+                                           type_schema['derived_from'],
```
Comment (simon-bar) [inline dsl_parser/elements/data_types.py]: Maybe do sth like derived_from = type_schema[_DERIVED_FROM]

File: dsl_parser/elements/data_types.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        types_internal = {}
+        for type_name, type_schema in datatypes.iteritems():
+            if type_name in constants.PRIMITIVE_TYPES:
+                raise exceptions.DSLParsingFormatException(
+                    1,
+                    "Illegal type name '{0}' - it is primitive "
+                    "type".format(type_name))
+
+            class DataTypeInternal(DataType):
+                requires = {}
```
Comment (simon-bar) [inline dsl_parser/elements/data_types.py]: and this line as well?

Comment (simon-bar) [inline dsl_parser/elements/data_types.py]: actually both requires and _type_name seem to be not necessary

File: dsl_parser/constants.py [hunk 6]
```diff
@@ -63,3 +63,5 @@
 IMPORT_RESOLVER_KEY = 'import_resolver'
 RESOLVER_IMPLEMENTATION_KEY = 'implementation'
 RESLOVER_PARAMETERS_KEY = 'parameters'
+
+PRIMITIVE_TYPES = ['string', 'integer', 'float', 'boolean']
```
Comment (dankilman) [inline dsl_parser/constants.py]: put next to elements constant and rename to USER_PRIMITIVE_TYPES

File: dsl_parser/elements/properties_utils.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    schema = Leaf(type=elements.PRIMITIVE_TYPES)
+
+
+class SchemaPropertyDescription(Element):
+
+    schema = Leaf(type=str)
+
+
+class SchemaPropertyType(Element):
```
Comment (dankilman) [inline dsl_parser/elements/properties_utils.py]: Restore validations for non data_type schemas to property type element

File: dsl_parser/utils.py [hunk 8]
```diff
@@ -40,12 +42,20 @@ def flatten_schema(schema):
     return flattened_schema_props
 
 
+def _property_description(path, name):
+    return '{0}.{1}'.format('.'.join(path), name)
```
Comment (dankilman) [inline dsl_parser/utils.py]: handle the case where not path at all so there will be not . prefix

File: dsl_parser/tests/test_data_types.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+                type: agent
+            agent_name:
+                type: string
+node_templates:
+    vm:
+        type: vm_type
+        properties:
+            agent: {}
+            agent_name: { get_property: [SELF, agent, connection, username] }
+data_types:
```
Comment (dankilman) [inline dsl_parser/tests/test_data_types.py]: add test for nested default merging values (some use defaults some don't)

Comment (dankilman) [inline dsl_parser/tests/test_data_types.py]: add test the subtype can override a property data_type with its own

File: dsl_parser/utils.py [hunk 10]
```diff
@@ -56,59 +66,124 @@ def merge_schema_and_instance_properties(
...
+                1,
+                "Illegal type name '{0}'".format(type_name))
+        val_clone = copy.deepcopy(property_schema)
+        default_value = property_schema.get('default')
+        if default_value:
+            undefined_property_error = 'Undefined property {1} in default' \
+                                       ' value of type {0}'
+            missing_property_error = 'Property {1} is missing in default' \
+                                     ' value of type {0}'
+            default_value = _parse_value(
```
Comment (dankilman) [inline dsl_parser/utils.py]: possible bug with derived data_types missing default value

File: dsl_parser/elements/data_types.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+_DERIVED_FROM = 'derived_from'
+_TYPE = 'type'
+
+
+class DataType(Element):
+    schema = {
+        constants.PROPERTIES: properties_utils.UnsafeSchema,
+        'description': StringElement,
+        _DERIVED_FROM: StringElement,
```
Comment (dankilman) [inline dsl_parser/elements/data_types.py]: see if you can reuse `DerivedFrom` in types.py

File: dsl_parser/elements/data_types.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        return result
+
+
+def _add_requirement_if_exists(current_type,
+                               required_type,
+                               types_classes,
+                               unknown_type_err):
+    if required_type not in types_classes:
+        raise exceptions.DSLParsingElementMatchException(
+            39,
```
Comment (dankilman) [inline dsl_parser/elements/data_types.py]: Verify error code is not used anywhere else and move as constant to dsl_parser.exceptions

File: dsl_parser/elements/data_types.py [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+
+    schema = Dict(type=SchemaProperty)
+
+
+class DataType(types.Type):
+
+    schema = {
+        constants.PROPERTIES: Schema,
+        'description': StringElement,
+        constants.DERIVED_FROM: types.DerivedFrom,
```
Comment (dankilman) [inline dsl_parser/elements/data_types.py]: I think this should be some class the derives from `types.DerivedFrom` which overrides the 
`descriptor`
https://github.com/cloudify-cosmo/cloudify-dsl-parser/blob/master/dsl_parser/elements/types.py#L58
attribute

File: dsl_parser/elements/data_types.py [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+    def parse(self, type_name, component_types):
+        if self.initial_value is None:
+            return
+        component_types = component_types or {}
+        prop_name = self.ancestor(SchemaProperty).name
+        undefined_property_error = 'Undefined property {1} in default' \
+                                   ' value of type {0}'
+        missing_property_error = 'Property {1} is missing in default' \
+                                 ' value of type {0}'
+        current_type = self.ancestor(Schema).parent().name
```
Comment (dankilman) [inline dsl_parser/elements/data_types.py]: Why not simply `self.ancestor(DataType).name`?

File: dsl_parser/elements/data_types.py [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+        if self.initial_value and self.initial_value not in \
+                constants.USER_PRIMITIVE_TYPES and not data_type:
+            raise exceptions.DSLParsingLogicException(
+                exceptions.ERROR_UNKNOWN_TYPE,
+                "Illegal type name '{0}'".format(self.initial_value))
+
+    def calculate_provided(self, data_type, component_types):
+        component_types = component_types or {}
+        if self.value and self.value not in constants.USER_PRIMITIVE_TYPES:
+            component_types = copy.copy(component_types)
```
Comment (dankilman) [inline dsl_parser/elements/data_types.py]: no need for `copy` as all the values injected are already copies of the original
https://github.com/simon-bar/cloudify-dsl-parser/blob/dsl_parser_datatypes/dsl_parser/framework/elements.py#L107
https://github.com/simon-bar/cloudify-dsl-parser/blob/dsl_parser_datatypes/dsl_parser/framework/elements.py#L122

https://github.com/simon-bar/cloudify-dsl-parser/blob/dsl_parser_datatypes/dsl_parser/framework/parser.py#L371
https://github.com/simon-bar/cloudify-dsl-parser/blob/dsl_parser_datatypes/dsl_parser/framework/parser.py#L387

File: dsl_parser/elements/data_types.py [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+            component_types[self.value] = data_type
```
Comment (dankilman) [inline dsl_parser/elements/data_types.py]: from 
https://github.com/cloudify-cosmo/cloudify-dsl-parser/pull/56/files#diff-495eb61eaa4c9fb6bc2700ae4d9da45eR200
and
https://github.com/cloudify-cosmo/cloudify-dsl-parser/pull/56/files#diff-495eb61eaa4c9fb6bc2700ae4d9da45eR163
it seems this line is redundant no?
as `data_type` will already be included in `component_types`

**Comment by barakm** [comment]
@dankilman, @konradxyz what is the status of this PR?

**Comment by konradxyz** [comment]
It's waiting for code review.

**Comment by dankilman** [comment]
work continued on a different branch

**Pull request closed without merging**